### PR TITLE
Organized Backbone Views. Edit Group works. Autopopulate works. Group lists in sync

### PR DESCRIPTION
**Pull request by BSteephenson**
This is a realllly big PR.
** This does NOT break anything (anything that wasn't broken before)

# Aims
- Organize front end so it's somewhat more manageable
- Have things sync across the page

# New Organization
- Renamed EventsView to EventsPageView, which better explains what that thing is doing
- EventsPageView has a UserView (represents the top bar thing)
- UserView has the buttons and the respective modals (contained in GroupViews.coffee and EventCRUDViews.coffee)
- GroupViews.coffee currently has AddGroupView and EditGroupView
- EventCRUDViews.coffee currently has AddEventView, EditEventView, and DeleteEventView

To picture it as an hierarchy 
- EventsPageView
    - UserView
        - AddGroupView
        - EditGroupView
        - AddEventView
        - EditEventView
        - DeleteGroupView
        - ( more to come as we do ChangeAdmins and DeleteGroup)

# Keeping Things In Sync
- EventsPageView has a UserModel
- EventsPageView passes this UserModel down to UserView
- UserView passes this UserModel down to the modals
- Since JS is pass by pointer for objects, everyone has a pointer to a singleton UserModel
- Each view can listen to events triggered by the UserModel and do the necessary DOM changes (example: @model.on 'groups.changed', @rerenderListOrSomething
- Each view makes form submissions through the UserModel

AddViewModel submits a form => UserModel makes the API call => UserModel refreshes => UserModel triggers 'groups.changed' => Everything on the page that has a list of groups makes the necessary updates

# What Works?
- Logging in and signing up
- Adding a group
- Adding an event
- Editing a group (plus auto-populate and the list is in sync) *this didn't work before*
- The Events listing stays in sync with edits to a group *this didn't work before*

# What Doesn't Work?
- Admin and Deleting a group (laziness, do it later,  I think Lenail is doing the admin part)
- Changing and Deleting an event (right now there's no straightforward way to get a user's group's events.. Need some API developments)

This isn't the perfect way to approach this, but I think this is a good base for making incremental fixes and changes in the future.

### CODE REVIEW

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+		if endTime.isBefore(startTime) then endTime.add(1, 'days')
+
+		Event.set 'startTime', startTime.toDate()
+		Event.set 'endTime', endTime.toDate()
+
+		@model.createEvent Event.toJSON(), (res) => 
+			$('.tl-addEventModal').modal('hide')
+			$('.tl-addEventModal').find('input,textarea').val('')
+			# notice says "event successfuly created"
+		Event.destroy() #this might be a bad idea in the future
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: I left myself a flag here forevers ago. Do you mind resolving this before we merge? If creating a new EventModel, serializing it, and destroying it is actually the appropriate action we want to take, then that's fine. I have no idea how this is done in real life...

File: tuftslife/assets/js/GroupViews.coffee [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	
+	events : 
+		"click button.tl-deleteGroup-button " : "attemptSubmitForm"
+		"change .tl-chgGroup-group" : "selectedGroupChanged"
+
+	initialize: ->
+		@renderFilter()
+		@delegateEvents()
+		@model.on
+			"groups.changed": () => @renderGroupNames()
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: one liner

File: tuftslife/assets/js/GroupViews.coffee [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+class EditGroupView extends Backbone.View
+	
+	events : 
+		"click button.tl-deleteGroup-button " : "attemptSubmitForm"
+		"change .tl-chgGroup-group" : "selectedGroupChanged"
+
+	initialize: ->
+		@renderFilter()
+		@delegateEvents()
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: what is @delegateEvents? it isn't in this class...

File: tuftslife/assets/js/GroupViews.coffee [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+			selectedClass: 'tl-selected-chgGroupFilters'
+			OneAtATime: true
+			defaultState: false
+
+	renderGroupNames : =>
+		groups = @model.get('groups')
+		names = _.pluck(groups, 'name')
+		@$('.tl-chgGroup-group').html('')
+		for name in names
+			console.log name
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: console logs.

File: tuftslife/assets/js/GroupViews.coffee [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+			@$('.tl-chgGroup-group').append("<option>#{name}</option>")
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: there's probably a more concise way to do 'renderGroupNames'. In fact I think you do this very concisely above, in the EventsView (modals) class

File: tuftslife/assets/js/GroupViews.coffee [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+		
+		#autopopulate the first one
+		if groups.length > 0
+			@selectedGroup = groups[0]
+			@populateFields(@selectedGroup)
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: this little block confuses me: 
- If the number of groups is 0, will the even be able to open the 'edit groups' modal? 
- If the number of groups is 0, is @selectedGroup null?

File: tuftslife/assets/js/GroupViews.coffee [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+	populateFields : (group) ->
+		@$('.tl-chgGroup-name').val(group.name)
+		@$('.tl-chgGroup-desc').val(group.description)
+		# $('.tl-chgGroupFilters').children intersect group.tags addclass selected
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: This is going to be slightly difficult to merge with the filters edits I made, I think. I changed this commented line to a functional one in my filters PR, which is now in master. Checkout the changed I made #22 to EventsView.

File: tuftslife/assets/js/GroupViews.coffee [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+	selectedGroupChanged: (e) =>
+		groupName = $(e.target).val()
+		@selectedGroup = @model.getGroupWithName(groupName)
+		console.log groupName
+		console.log @selectedGroup.id
+		console.log @model.get('groups')
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: I imagine we won't need these logs?

File: tuftslife/assets/js/UserModel.coffee [hunk 9]
```diff
@@ -1,19 +1,33 @@
...
+# 	creating/editing events
+
 class UserModel extends Backbone.Model
 	
 	initialize : ->
 		@set('logged.in', false) 
-		@fetchUser()
-
 
+	# Basically a refresh function. Called (manually) after every API call
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: This seems a little bit dangerous. Can you justify this? I'm not sure how many calls we can handle.

File: tuftslife/assets/js/UserModel.coffee [hunk 10]
```diff
@@ -1,19 +1,33 @@
...
-				@trigger 'logged.in'
+
+				# Just refresh events and group every time
+				@trigger 'groups.changed'
+				@trigger 'my.events.changed'
+
 			cb?()
 
+	# An alias cuz it's shorter and makes sense in the code
+	refresh: () => @fetchUser()
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: Yeah, I can imagine we could also need to refresh other things in the future as well. good thinking.

File: tuftslife/assets/js/UserModel.coffee [hunk 11]
```diff
@@ -36,11 +50,14 @@ class UserModel extends Backbone.Model
 			@trigger('logged.out')
 			@fetchUser () => cb?(res)
 
+	# Gets full name from UTLN
+	getName: () -> 
+		(@get('username').split('.').map (s) -> s.charAt(0).toUpperCase() + s.slice(1)).join(' ')
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: Why was this added up here and removed from down there? Also, no need for parens here. Functions are defined entirely by arrows, and if they have no arguments, they don't need parens

File: tuftslife/assets/js/UserModel.coffee [hunk 12]
```diff
@@ -48,15 +65,36 @@ class UserModel extends Backbone.Model
...
-	getName: -> (@get('username').split('.').map (s) -> s.charAt(0).toUpperCase() + s.slice(1)).join(' ')
+	editGroup: (group, cb) ->
+		return cb("No group id") if !group.id?
+		$.post '/group/edit/' + group.id, group, (res) =>
+			#start refreshing groups
+			@refresh()
+			return cb?(res)
 
+	editGroupAdmin: (groupname, action, username, cb) ->
+		return cb("missing field") if !(groupname? && action? && username?)
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: in coffee, I'd rather you use 'and' instead of '&&'. Also, are you sure you're DeMorgan'd here properly?

File: tuftslife/assets/js/UserView.coffee [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+		@EditEventView = new EditEventView
+			el: $('.tl-chgEventModal')
+			model: @model
+
+
+	render: () ->
+		if @model.get('logged.in')
+			@setupAdminTools()
+		else
+			@renderLogin()
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: I personally prefer

+		if @model.get('logged.in') then @setupAdminTools() else @renderLogin()

File: tuftslife/assets/js/UserView.coffee [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+	# helper to make creating popovers easier
+	popOptions:
+		trigger: 'click'
+		html: true
+		template: '<div class="popover" role="tooltip"><div class="arrow"></div><div class="popover-content"></div><h3 class="popover-title"></h3></div>'
+		placement: 'bottom'
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: I guess I don't like this object being a property of the view. I wish it were more hidden in a way, like an object in one of these functions here.

File: tuftslife/assets/js/UserView.coffee [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+	switchPopoverContent: (title, content, cb) =>
+		@$el.popover('hide')
+		@$el.popover('destroy')
+		
+		@popOptions.title = title
+		@popOptions.content = content
+
+		popover = @$el.popover(@popOptions)
+
+		#@$el.popover('show')
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: ?

File: tuftslife/assets/js/UserView.coffee [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+
+		popover.on 'shown.bs.popover', =>
+			cb?()
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: one-line

File: tuftslife/assets/js/UserView.coffee [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+
+		@model.login utln, password, (res) => 
+
+			if res.message?
+				if res.message is 'Unable to authenticate'
+					message = 'We can\'t find an account with those credentials. Have you signed up?'
+				if res.message is 'Account unactivated'
+					message = 'Your account seems to be unactivated. Check your email for activation instructions.'
+				if res.message is 'Unable to login'
+					message = 'Req.login failed. Check console logs'
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: I've always sort of been ashamed of this piece of code. We can fix it later, but if you have an urge to fix it now, go for it.

File: tuftslife/assets/js/UserView.coffee [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+								
+				
+				@switchPopoverContent "Login Failed", message
+
+				hidePopover = => $(@popover).popover('hide'); @renderLogin()
+				window.setTimeout(hidePopover, 4000)
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: This is the real shameful part though. It's a bug. No need to fix it here, but I'll fix it immediately after we merge.

File: tuftslife/assets/js/UserView.coffee [hunk 19]
```diff
@@ -0,0 +1,131 @@
...
+				@switchPopoverContent "Login Failed", message
+
+				hidePopover = => $(@popover).popover('hide'); @renderLogin()
+				window.setTimeout(hidePopover, 4000)	
+
+
+			else
+				@$el.popover('destroy')
+				@setupAdminTools()
+				@model.on 'groups.changed', => @setupAdminTools()
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: you're binding an event here. that makes me nervous. I though you had ported all of this functionality somewhere else?

File: tuftslife/tasks/pipeline.js [hunk 20]
```diff
@@ -57,7 +57,9 @@ var jsFilesToInject = [
   'js/CalendarModel.js',
   'js/CalendarView.js',
 
-
+  'js/UserModel.js',
+  'js/GroupViews.js',
+  'js/UserView.js',
```
Comment (zfrenchee) [inline tuftslife/tasks/pipeline.js]: EventCrud? We should also rename that file. EventViews I think will do. 
Also EventPageView

File: tuftslife/assets/js/UserView.coffee [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+		$popover = $('.popover')
+		utln = $popover.find('.tl-utln-field').val()
+		passwords = $popover.find('.tl-signup-password').map -> $(@).val()
+		if passwords[0] isnt passwords[1] then alert 'passwords dont match'; return
+		if passwords[0].length < 6 then alert 'password must be a minimum of 6 characters'; return
+
+		@model.signUp(utln, passwords[0], (res) =>
+			message = "Welcome. Check your email to complete your registration."
+
+			@switchPopoverContent('Welcome to Tuftslife!', message)
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: There are a few extra lines in here you can get rid of for clarity.

File: tuftslife/assets/js/UserView.coffee [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+
+	# Shows the logout button, the groups button, the events button when appropriate
+	setupAdminTools: () =>
+		$('.tl-login-button').text('Logout').click => @logout()
+		$('.tl-nav-message').text("Logged in as #{@model.getName()}")
+		$('.tl-nav-message, .tl-group-dropdown').removeClass('hidden')
+		if @model.get('groups')? and @model.get('groups').length > 0 
+			$('.tl-event-dropdown').removeClass('hidden')
+			@AddEventView.render()
+			@EditEventView.render()
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserView.coffee]: I feel like we shouldn't be calling render from outside of these classes. They should probably be listening to some state, and render themselves. This is precarious.

File: tuftslife/assets/js/EventsPageView.coffee [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+# Represents the Events page
+# Has all views dealing with getting, creating, editing, deleting Events
+# Has a UserModel which has all groups and events and does all AJAX calls
+
+class EventsPageView extends Backbone.View
+	
+	initialize: ->
+		@User = new UserModel()		
+		@User.on
+			"groups.changed": @constructEvents
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventsPageView.coffee]: This isn't hugely clear: why would we update the events when the groups change?

File: tuftslife/assets/js/EventsPageView.coffee [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+		@Events = new EventsCollection
+
+		# Doesn't render on init
+		@UserView = new UserView
+			el: $('.tl-login-button')
+			model: @User
+
+		@User.fetchUser()
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventsPageView.coffee]: This should be in the initialization of user, I feel like.

File: tuftslife/assets/js/EventsPageView.coffee [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+			$('.tl-eventInfo-place').text Event.get('place')
+			$('.tl-eventInfo-groupName').text Group.name
+			$('.tl-eventInfo-groupDesc').text Group.description
+
+			$(".tl-event").each -> $(@).removeClass('tl-event-selected')
+			$(".tl-event##{eventID}").addClass('tl-event-selected')
+
+events = new EventsPageView
+
+
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventsPageView.coffee]: I feel like there must be a better way to do this than all that jquery =((((((

**Comment by zfrenchee** [comment]
I have a comment that might upset you, @BSteephenson: Do you think you could pull UserView

**Comment by BSteephenson** [comment]
I
- made it so you only render EventsPage after (via callback) User Model fetches the user. Simplifies things a lot imo
- put back the whole Login popover code back to the way it was.

I did not
- attempt to make the tags thing work in Edit Group. See the other closed PR

**Comment by BSteephenson** [comment]
And things render themselves. It all works

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+
+	render: =>
+		$('.tl-addEvent-from').timepicker()
+		$('.tl-addEvent-to').timepicker()
+
+		@Calendar = new CalendarView 
+			el: $('.tl-addEvent-calendar')
+			cb: (date) -> null
+
+		$('.tl-addEvent-submit-btn').click => @submitForm()
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: you could put this under 'events' (where you currently have a #TODO). That would conform better with the style you have been employing.

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 27]
```diff
@@ -0,0 +1,160 @@
+
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: Can we rename this file from CRUD?

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+	render: ->
+		$('.tl-chgEvent-from').timepicker()
+		$('.tl-chgEvent-to').timepicker()
+
+		@Calendar = new CalendarView 
+			el: $('.tl-chgEvent-calendar')
+			cb: (date) -> null
+
+		$('.tl-chgEvent-submit-btn').click => @submitForm()
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: we can also put this under 'events' I think.

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+
+		@renderEventsList(selectedGroup)
+
+	renderEventsList: (group) =>
+
+		$('.tl-chgEvent-event').html('')
+		if group?
+			for event in @group.events
+				$('.tl-chgEvent-event')
+				.append("<option>#{event.name}</option>")
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: put these on the same line?

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+			$('.tl-chgEvent-event').click(@selectedEventChanged)
+		else
+			# Tell user to pick a group first
+			$('.tl-chgEvent-event')
+			.append("<option disabled>Select a Group</option>")
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: you can also one line these

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 31]
```diff
@@ -0,0 +1,160 @@
...
+	populateFields: (event) ->
+		$('.tl-chgEvent-name').val(event.get('name'))
+		$('.tl-chgEvent-description').val(event.get('description'))
+		$('.tl-chgEvent-place').val(event.get('place'))
+		
+		startTime = moment(event.get('startTime'))
+		endTime = moment(event.get('endTime'))
+
+		$('.tl-chgEvent-from').val(startTime.format('h:mm A'))
+		$('.tl-chgEvent-from').val(endTime.format('h:mm A'))
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: '.tl-chgEvent-to'

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 32]
```diff
@@ -0,0 +1,160 @@
...
+
+		Event.set 'startTime', startTime.toDate()
+		Event.set 'endTime', endTime.toDate()
+
+		@model.editEvent Event.toJSON(), (res) => 
+			$('.tl-chgEventModal').modal('hide')
+			$('.tl-chgEventModal').find('input,textarea').val('')
+			# notice says "event successfuly created"
+		Event.destroy() #this might be a bad idea in the future
+		@$el.modal('hide')
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: Does this work with the backend? if you post this to editEvent, does it edit the right event?

File: tuftslife/assets/js/EventCRUDViews.coffee [hunk 33]
```diff
@@ -0,0 +1,160 @@
...
+	# receives an el and a model from parent
+
+	events: null # TODO
+
+	initialize: -> # add listeners to @model
+
+	submitForm: ->
+
+	renderGroupList: ->
+
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventCRUDViews.coffee]: Put 'null's at the end of each arrow. i.e. initialize: -> null

File: tuftslife/assets/js/EventsPageView.coffee [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+		@User = new UserModel
+			# Only do render AFTER the user has been fetched
+			afterFetch: @render
+
+		# Refresh the events list and content whenever groups change	
+		@User.on "groups.changed", @constructEvents
+
+		@Events = new EventsCollection
+
+		@render()
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventsPageView.coffee]: Are you calling render twice here? UserModel is going to fetch user, then render. Also you're calling render again. Do you need to?

File: tuftslife/assets/js/EventsPageView.coffee [hunk 35]
```diff
@@ -0,0 +1,98 @@
...
+	initialize: ->
+
+		@User = new UserModel
+			# Only do render AFTER the user has been fetched
+			afterFetch: @render
+
+		# Refresh the events list and content whenever groups change	
+		@User.on "groups.changed", @constructEvents
+
+		@Events = new EventsCollection
```
Comment (zfrenchee) [inline tuftslife/assets/js/EventsPageView.coffee]: I would personally switch the ordering of the above two lines, I think. Semantically clearer, right?
- here's a user
- here's a bunch of events
- if the user changes the groups, update the events. 
seems like the best ordering of those statements.

File: tuftslife/assets/js/GroupViews.coffee [hunk 36]
```diff
@@ -0,0 +1,86 @@
...
+
+class AddGroupView extends Backbone.View
+	
+	events : 
+		"click button.tl-addGroup-submit-btn" : "attemptSubmitForm"
+
+	initialize: ->
+		@renderFilter()
+
+	renderFilter: ->
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: let's call this method 'render' even though right now it only renders filters.

File: tuftslife/assets/js/GroupViews.coffee [hunk 37]
```diff
@@ -0,0 +1,86 @@
...
+
+	initialize: ->
+		@model.on "groups.changed", @constructGroupsList
+		@render()
+
+	render: ->
+		@renderFilter()
+		if @model.get('groups')? then @constructGroupsList()	
+
+	renderFilter: ->
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: just put this in render imo. No need for a separate function, right?

File: tuftslife/assets/js/GroupViews.coffee [hunk 38]
```diff
@@ -0,0 +1,86 @@
...
+		#autopopulate the first one
+		if @model.get('groups').length > 0
+			@selectedGroup = @model.get('groups')[0]
+			@populateFields(@selectedGroup)
+
+	populateFields : (group) ->
+		@$('.tl-chgGroup-name').val(group.name)
+		@$('.tl-chgGroup-desc').val(group.description)
+		# TODO???
+		#@GroupTags.selectTags(group.tags)
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: as soon as we merge your PR, i'll open one which fixes this.

File: tuftslife/assets/js/GroupViews.coffee [hunk 39]
```diff
@@ -0,0 +1,86 @@
...
+
+	attemptSubmitForm: () ->
+
+		@model.editGroup
+			id: @selectedGroup.id
+			name: $('.tl-chgGroup-name').val()
+			description: $('.tl-chgGroup-desc').val()
+			tags: @GroupTags.active()
+		@$el.modal('hide')
+
```
Comment (zfrenchee) [inline tuftslife/assets/js/GroupViews.coffee]: So it looks like this class can only handle the 'edit' part of the modal right now. Just wanted to clarify that that was your intention. Delete *does* have a frontend right now, but it isn't necessarily complete.

File: tuftslife/assets/js/UserModel.coffee [hunk 40]
```diff
@@ -1,19 +1,36 @@
...
+
 class UserModel extends Backbone.Model
 	
-	initialize : ->
-		@set('logged.in', false) 
-		@fetchUser()
+	initialize: (params) ->
 
+		@set('logged.in', false)
+		@fetchUser(params.afterFetch)
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: Should this be a callback or a backbone event that someone can listen to. I think you're right that a callback is best here, but I can't say for sure. What are the pros/cons?

File: tuftslife/assets/js/UserModel.coffee [hunk 41]
```diff
@@ -1,19 +1,36 @@
...
 
+		@set('logged.in', false)
+		@fetchUser(params.afterFetch) 
 
 	fetchUser: (cb) ->
 		$.get '/user', (res) => 
 			if res.user?
 				@set('username', res.user.username)
 				@set('groups', res.user.groups)
+				if @get('logged.in') is false then @trigger 'logged.in groups.changed my.events.changed'
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: I haven't seen 'my.events.changed' used yet. Is that an event that's going to be used later or did you forget to take it out?

File: tuftslife/assets/js/UserModel.coffee [hunk 42]
```diff
@@ -48,15 +67,36 @@ class UserModel extends Backbone.Model
 	getGroupWithName: (name) -> _.find(@get('groups'), (element) -> element.name is name)
 
 
-	editGroup: (group) -> null # TODO
-
-
-	getName: -> (@get('username').split('.').map (s) -> s.charAt(0).toUpperCase() + s.slice(1)).join(' ')
+	editGroup: (group, cb) ->
+		return cb("No group id") if !group.id?
```
Comment (zfrenchee) [inline tuftslife/assets/js/UserModel.coffee]: I'd prefer if you used 'not' instead of '!' and 'if (...) then return (...)' rather than 'return (...) if (...)' here (and also in the next function). What do you think?

**Comment by zfrenchee** [comment]
Sweet. It's nearly ready to go. I'm really excited to merge this PR!

File: tuftslife/assets/js/EventCRUDViews.coffee (hunk 31 above)
Comment (BSteephenson) [inline tuftslife/assets/js/EventCRUDViews.coffee]: That's an oops.
Should I somehow mark in the code that this part (editing events) is still not done / not tested?

File: tuftslife/assets/js/EventCRUDViews.coffee (hunk 32 above)
Comment (BSteephenson) [inline tuftslife/assets/js/EventCRUDViews.coffee]: That was copy paste. This whole part is untested as we don't have all the backend functionality for it (we have an edit method, we just don't know what events a user's groups' have)

File: tuftslife/assets/js/EventsPageView.coffee (hunk 34 above)
Comment (BSteephenson) [inline tuftslife/assets/js/EventsPageView.coffee]: oops

File: tuftslife/assets/js/UserModel.coffee (hunk 40 above)
Comment (BSteephenson) [inline tuftslife/assets/js/UserModel.coffee]: I think a callback for the first initializing is best. Reduces complexity to it.
Think of the timeline as this
fetch User => render stuff => wait for user to do something
Limiting the events pattern to that last part (UI interaction) makes life easier imo

File: tuftslife/assets/js/UserModel.coffee (hunk 41 above)
Comment (BSteephenson) [inline tuftslife/assets/js/UserModel.coffee]: TODO for the future. Ideally, UserModel would be able to figure out what changed and emit the right events. That way you won't need to do something like Update the groups list whenever you Make an event. Right now, it's naively just updating everything.

File: tuftslife/assets/js/GroupViews.coffee (hunk 39 above)
Comment (BSteephenson) [inline tuftslife/assets/js/GroupViews.coffee]: TODO added to the trello

**Comment by BSteephenson** [comment]
I
- used Backbone.View's events rather than jQuery
- made things one liners where you asked me to
- fixed the DOM class .tl-chg-Event-to thingy
- put comments indicating that EditEvent is not working and is not tested
- got rid of the second render() call in EventsPageView
- reorder method calls in EventsPageView's initialize
- gave everything a render, put renderFilter's code inside there in places
- used not instead of ! and if ... then ... instead of .... if .....
- renamed EventCrudViews to EventViews.coffee

**Comment by zfrenchee** [comment]
YURSS!
